service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -1,93 +1,119 @@
+    public CertificateDTO update(CertificateDTO dto) {
+        Certificate existing = certificateDAO.readById(dto.getId())
+                .orElseThrow(() -> new IdNotExistException(dto.getId().toString()));
+        updateFieldsFromDto(existing, dto);
+        existing = certificateDAO.update(existing).orElseThrow(UpdateEntityInternalException::new);
+        return modelMapper.map(existing, CertificateDTO.class);
     }
 
     @Override
-    public List<CertificateDTO> searchByParameters(String tagName, String certificateName, String description, String sortBy, String order) {
-        String queryPart = queryProvider.getCertificateSearchQuery(tagName, certificateName, description, sortBy, order);
-        List<String> queryParameters = queryParametersProvider.getCertificateParameters(tagName, certificateName, description);
-
-        return certificateDAO.searchByParameters(queryPart, queryParameters)
-                .stream()
-                .map(certificate -> certificateDTOMapper.toDto(certificate, tagService.readByCertificateId(certificate.getId())))
+    public List<CertificateDTO> searchByParameters(SearchParameter searchParameter, int page, int size) {
+        List<Certificate> certificates = certificateDAO.searchByParameters(searchParameter, page, size);
+        return certificates.stream()
+                .map(certificate -> modelMapper.map(certificate, CertificateDTO.class))
                 .collect(Collectors.toList());
     }
 
     @Override
     @Transactional
-    public Long create(CertificateDTO dto) {
-        // insert certificate
-        Certificate certificateEntity = certificateDTOMapper.toEntity(dto);
-        Long certificateId = certificateDAO.create(certificateEntity);
-
-        // insert tags
-        List<TagDTO> tagsDTO = dto.getTags();
-        tagsDTO.forEach(tagDTO -> {
-            Long tagId = tagService.readByName(tagDTO.getName()).map(Tag::getId).orElseGet(() -> tagService.create(tagDTO));
-            certificateDAO.createCertificateHasTag(certificateId, tagId);
-        });
-        return certificateId;
+    public CertificateDTO create(CertificateDTO dto) {
+        Certificate certificate = modelMapper.map(dto, Certificate.class);
+        // get tags or create new ones
+        List<Tag> tags = dto.getTags().stream()
+                .map(tagDTO ->
+                        tagDao.readByName(tagDTO.getName())
+                                .orElseGet(() -> tagDao.create(modelMapper.map(tagDTO, Tag.class))
+                                        .orElseThrow(CreateEntityInternalException::new))
+                )
+                .collect(Collectors.toList());
+        certificate.setTags(tags);
+
+        // create certificate
+        Certificate created = certificateDAO.create(certificate)
+                .orElseThrow(CreateEntityInternalException::new);
+        return modelMapper.map(created, CertificateDTO.class);

```

Comment:
this method makes at least 2 operations with different entities, by this reason divide into 2 methods

Reply:
the same in other classes

Reply:
thanks for pointing this. method was divided